flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/catalog/FlinkCatalogManager.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.catalog;
+
+import org.apache.flink.table.api.CatalogAlreadyExistsException;
+import org.apache.flink.table.api.CatalogNotExistException;
+import org.apache.flink.util.StringUtils;
+
+import org.apache.calcite.jdbc.CalciteSchema;
+import org.apache.calcite.schema.SchemaPlus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * A CatalogManager implementation for Flink.
+ * TODO: [FLINK-11275] Decouple CatalogManager with Calcite
+ *   Idealy FlinkCatalogManager should be in flink-table-api-java module.
+ *   But due to that it currently depends on Calcite, a dependency that flink-table-api-java doesn't have right now.
+ *   We temporarily put FlinkCatalogManager in flink-table-planner-blink.
+ */
+public class FlinkCatalogManager implements CatalogManager {
+	private static final Logger LOG = LoggerFactory.getLogger(FlinkCatalogManager.class);
+
+	public static final String BUILTIN_CATALOG_NAME = "builtin";
+
+	// The catalog to hold all registered and translated tables
+	// We disable caching here to prevent side effects
+	private CalciteSchema internalSchema = CalciteSchema.createRootSchema(false, false);
+	private SchemaPlus rootSchema = internalSchema.plus();
+
+	// A map between names and catalogs.
+	private Map<String, ReadableCatalog> catalogs;
+
+	// The name of the default catalog and schema
+	private String currentCatalogName;
+
+	public FlinkCatalogManager() {
+		LOG.info("Initializing FlinkCatalogManager");
+		catalogs = new HashMap<>();
+
+		GenericInMemoryCatalog inMemoryCatalog = new GenericInMemoryCatalog(BUILTIN_CATALOG_NAME);
```

Comment:
Good question, this PR only partially reflects the integration of `CatalogManager` and `TableEnv`. We propose adding new APIs to `TableEnv` to read/write catalog metadata. If users try to write to a `ReadableCatalog`, a catalog exception will be thrown to remind them that. E.g. for CatalogBaseTable, we'll have

```
// can have overloaded methods or a different signature to register table in a catalog different than the current one
void registerTable(String name, CatalogBaseTable table, boolean ignoreIfExists) {
   if (!(catalogManager.getCurrentCatalog() instanceof ReadableWritableCatalog)) {
     throw CatalogException("xxx is a ReadableCatalog only")
   } 
   catalog = (ReadableWritableCatalog) catalogManager.getCurrentCatalog()
   catalog.createTable(new ObjectPath(catalog.getCurrentDb(), name), table, ignoreIfExists)
}
```

`ReadableCatalog` and `ReadableWritableCatalog` is inherited from the original `ExternalCatalog` and `CrudExternalCatalog`, more context in [FLIP-30](https://cwiki.apache.org/confluence/display/FLINK/FLIP-30%3A+Unified+Catalog+APIs).

I believe it's good to have `ReadableCatalog`. It doesn't make sense or just cannot support writing in some catalog impl. E.g. at least for now I don't see  Confluent Registry catalog ( [FLINK-12256](https://issues.apache.org/jira/browse/FLINK-12256)) needs to support writing since writing thru Confluent Registry itself and its fancy UI is much more convenient and user friendly.

Reply:
I agree that there might be implementations that for time being do not support writing and might throw exception. In general I think though all the implementation do support writing in a way. I agree e.g. the owner of a platform might decide to disallow writing, but this should be rather done with some sort of access roles or property base etc. In the end all catalogs do support writing, otherwise there would be no way to enter data into them.

The method you mentioned `registerTable` is part of the `ReadableWritableCatalog` not `ReadableCatalog`, if the `CatalogManager` works only with `ReadableCatalog`, we would have to perform instance checks and hard casts all the time. I see no reason for that. I believe we can just have a single `Catalog` interface that might throw exception if it should not be modified or just expose a flag such as `isMutable`.

Reply:
I think access control/authorization is orthogonal to this problem.

I'd argue that we probably should allow some types of catalog to be readable only, and we just don't enter data into them but only read data from them. With the example of Confluent Registry, it has really nice features such as fancy UI, schema evolution, compatibility check, etc. If users are already using that to manage their kafka schemas, why would they wanna use something else that don't support all these features or not natively? Support writing in Flink's Registry Catalog would be vain IMHO.

The `registerTable` is a part of `TableEnvironment` in the example. With the use case above in mind, having only a`Catalog` interface and letting a catalog implement 13+ methods (the # can grow) that it doesn't support seem to be not a good design.

I wonder whether we should add a `CatalogManager#getReadableWritableCatalog()` API, which can perform checks in side and `TableEnvironment` won't worry about checks anymore.